pkg/controllers/vdb/sandboxsubcluster_reconciler.go

```diff
@@ -0,0 +1,184 @@
+		VRec:       vdbrecon,
+		Log:        log.WithName("SandboxSubclusterReconciler"),
+		Vdb:        vdb,
+		PFacts:     pfacts,
+		Dispatcher: dispatcher,
+		Client:     cli,
+	}
+}
+
+// Reconcile will add subclusters to sandboxes if we found any qualified subclusters
+func (s *SandboxSubclusterReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	// no-op for ScheduleOnly init policy or enterprise db or no sandboxes
+	if s.Vdb.Spec.InitPolicy == vapi.CommunalInitPolicyScheduleOnly ||
+		!s.Vdb.IsEON() || len(s.Vdb.Spec.Sandboxes) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// We need to collect pod facts for finding qualified subclusters
+	if err := s.PFacts.Collect(ctx, s.Vdb); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return s.sandboxSubclusters(ctx)
+}
+
+// sandboxSubclusters will add subclusters to their sandboxes defined in the vdb
+func (s *SandboxSubclusterReconciler) sandboxSubclusters(ctx context.Context) (ctrl.Result, error) {
+	// find qualified subclusters with their sandboxes
+	scSbMap, err := s.fetchSubclustersWithSandboxes()
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if len(scSbMap) == 0 {
+		s.Log.Info("No subclusters need to be sandboxed")
+		return ctrl.Result{}, nil
+	}
+
+	// find an initiator to call vclusterOps
+	foundInitiator := false
+	for _, v := range s.PFacts.Detail {
+		// use an UP primary node in main cluster as the initiator
+		if !foundInitiator && v.sandbox == "" && v.isPrimary && v.upNode {
+			s.InitiatorIP = v.podIP
+			foundInitiator = true
+		}
+	}
+	if !foundInitiator {
+		return ctrl.Result{}, fmt.Errorf("cannot find an UP node in main cluster to execute sandbox operation")
+	}
+
+	// call sandbox API in vclusterOps, then update sandbox status in vdb
+	succeedSbScMap := make(map[string][]string)
+	for sc, sb := range scSbMap {
+		err := s.sandboxSubcluster(ctx, sc, sb)
+		if err != nil {
+			// when one subcluster failed to be sandboxed, update sandbox status and return error
+			return ctrl.Result{}, errors.Join(err, s.updateSandboxStatus(ctx, succeedSbScMap))
+		} else {
+			succeedSbScMap[sb] = append(succeedSbScMap[sb], sc)
+		}
+	}
+	return ctrl.Result{}, s.updateSandboxStatus(ctx, succeedSbScMap)
+}
+
+// fetchSubclustersWithSandboxes will return the qualified subclusters with their sandboxes
+func (s *SandboxSubclusterReconciler) fetchSubclustersWithSandboxes() (map[string]string, error) {
+	vdbScSbMap := s.Vdb.GenSubclusterSandboxMap()
+	targetScSbMap := make(map[string]string)
+	for _, v := range s.PFacts.Detail {
+		sb, ok := vdbScSbMap[v.subclusterName]
+		// skip the pod in the subcluster that doesn't need to be sandboxed
+		if !ok {
+			continue
+		}
+		// skip the pod in the subcluster that already in the target sandbox
+		if sb == v.sandbox {
+			continue
+		}
+		// skip the pod in the subcluster that is in another sandbox,
+		// the subcluster should be unsandboxed first by unsandbox reconciler
+		if v.sandbox != "" {
```

Comment:
We collect podfacts for either the main cluster or a sandbox. I think sandboxsubcluster  reconciler will be collecting the main cluster podfacts so it is fair to assume that sandboxed subclusters will have none of their pods in the podfacts and in that case checking up nodes should be enough.

Reply:
This check can be useful. Sooner or later, we should collect podfacts of all main clusters and sandboxes for this case: a user has a sandbox(name: sandbox1, subclusters: [sc1]) in vdb, and the user wants to change it to sandbox(name: sandbox1, subclusters: [sc2]).

Reply:
You  can only sandbox a subcluster that is part of the main cluster. The Pfacts that `sandboxsubcluster`  from vdb controller only collects pods in the main cluster(You can check podfacts initialization in vdb controller). As for your example I assume we would first need to unsandbox sc1 (separate reconciler) and then sandbox sc2.
My point is that that pfacts is already set to collect only main cluster pods.

Reply:
I see what you mean. In unsandbox reconciler, we will only collect pods for sandboxes. In sandbox reconciler, we only collect pods for main cluster. I will remove this check.